content/blog/typescript-type-generics/index.md

```diff
@@ -0,0 +1,340 @@
+	} catch (e) {
+		err = e;
+	}
+
+	return {
+		loggedValue: jsonString,
+		original: item,
+		// If there was no error, return `undefined` here
+		err: err
+	}
+}
+```
+
+If we wanted to type the `logTheValue` function, we'd want to make sure to use a type generic for the input parameter `item`. By doing so, we could use that same generic for the return prop of `loggedValue` to ensure they both have the same typing. To do this, we could do so inline:
+
+```typescript
+// Because this is an `async` function, we want to wrap the returned type value in a Promise
+async function logTheValue<ItemT>(item: ItemT): Promise<{loggedValue: string, original: ItemT, err: Error | undefined}> {
+	// ... Function body here
+}
+```
+
+Alternatively, we could utilize another feature of generics — the ability to pass the type value of the generic manually — and make an interface with a generic and do so there:
+
+```typescript
+interface LogTheValueReturnType<originalT> {
+  loggedValue: string;
+  original: originalT;
+  err: Error | undefined;
+}
+
+// Notice how we're even wrapping that interface in a built-in type with a generic argument for `Promise`!
+async function logTheValue<ItemT>(item: ItemT): Promise<LogTheValueReturnType<ItemT>> {
+	// ... Function body here
+}
+```
+
+WIth these few features, we're able to utilize much of the functionality of generics. 
+
+However, I know I haven't answered what the `<>` really is for. Well, much like type variables, there's also the ability to pass types as "type arguments" when generics are applied to a function.
+
+An example of this would be a syntax like this:
+
+```typescript
+logTheValue<number>(3);
+```
+
+# Non-Function Generics {#non-function-generics}
+
+As you saw before with the `LogTheValueReturnType` interface — functions aren't the only ones with generics. In addition to using them within functions and interfaces, you can also use them in classes. 
+
+Classes with generics can be particularly helpful for data structures like this:
+
+```typescript
+// DataType might want to be a base64 encoded string, a buffer, or an IntArray
+class ImageType<DataType> {
+	data: DataType;
+  height: number;
+  width: number;
+
+  constructor(data: DataType, height: number, width: number) {
+		this.data = data;
+		this.height = height;
+		this.width = width
+	};
+}
+
+function handleImageBuffer(img: ImageType<Buffer>) {}
+```
+
+Type generics in classes can be used as method argument and property types alike.
+
+There's also the ability to use generics within `type` definitions:
+
+```typescript
+interface ImageType<DataType> {
+	data: DataType;
+  height: number;
+  width: number;	
+}
+
+interface ImageConvertMethods<DataType> {
+  // This is the typing of a method. It will take a prop of the generic type and return the generic type
+  toPNG: (data: DataType) => DataType;
+  toJPG: (data: DataType) => DataType;
+}
+
+
+type ImageTypeWithConvertMethods<DataType> = ImageType<DataType> & ImageConvertMethods<DataType>
```

Comment:
I've noticed some inconsistent indenting in some of these code blocks. I don't have my editor up at the moment, but maybe some find and replace regex would fix this baby right up!

Reply:
Yeah, I'll be fixing the formatting of the code blocks in the next commit to use tabs as usual :)